sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,364 @@
+        // start of the new region is actually offset + delta, not offset
+        let xoff = if delta_x > 0 {
+            self.u_texture_offset.value.x
+        } else {
+            (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size)
+        };
+        let yoff = if delta_y > 0 {
+            self.u_texture_offset.value.y
+        } else {
+            (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size)
+        };
+
+        // u_texture_offset.value is always in [0; self.size)
+        let x = (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size);
+        let y = (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size);
+        self.u_texture_offset.value = Vector2::new(x, y);
+
+        let vert_regions = UpdateRegion::new_regions(xoff, y, self.size, &vert_strip);
+        let hori_regions = UpdateRegion::new_regions(x, yoff, self.size, &hori_strip);
+        let regions = ArrayVec::from(vert_regions)
+            .into_iter()
+            .chain(ArrayVec::from(hori_regions));
+
+        unsafe {
+            self.gl.BindTexture(opengl::TEXTURE_2D, self.id);
+            for r in regions {
+                let width = i32::try_from(r.pixels.width()).unwrap();
+                let height = i32::try_from(r.pixels.height()).unwrap();
+                if height * width == 0 {
+                    continue;
+                }
+                let image = r.pixels.to_image();
+                self.gl.TexSubImage2D(
+                    opengl::TEXTURE_2D,
+                    0,
+                    r.x,
+                    r.y,
+                    width,
+                    height,
+                    P::FORMAT,
+                    P::DTYPE,
+                    image.into_raw().as_ptr() as *const c_void,
+                );
+            }
+        }
+    }
+
+    /// Updates the offset and binds the texture
+    pub fn submit(&mut self) {
+        unsafe {
+            self.u_texture_offset.submit();
+
+            self.gl.ActiveTexture(opengl::TEXTURE0);
+            self.gl.BindTexture(opengl::TEXTURE_2D, self.id);
+        }
+    }
+}
+
+/// Trait to associate pixel formats with OpenGL constants
+pub trait TextureFormat: Pixel {
+    const INTERNALFORMAT: GLint;
+    const FORMAT: GLuint;
+    const DTYPE: GLuint;
+}
+
+impl TextureFormat for LumaA<f32> {
+    const INTERNALFORMAT: GLint = opengl::RG32F as GLint;
+    const FORMAT: GLuint = opengl::RG;
+    const DTYPE: GLuint = opengl::FLOAT;
+}
+
+impl TextureFormat for Rgba<u8> {
+    const INTERNALFORMAT: GLint = opengl::RGBA8 as GLint;
+    const FORMAT: GLuint = opengl::RGBA;
+    const DTYPE: GLuint = opengl::UNSIGNED_BYTE;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use image::GenericImage;
+
+    fn assert_same<I, J, P>(left: &I, right: &J)
+    where
+        I: GenericImageView<Pixel = P>,
+        J: GenericImageView<Pixel = P>,
+        P: Pixel + Eq,
+    {
```

Comment:
nit: I think this whole function could be replaced by `assert!(left.pixels.eq(right.pixels))`.